src/internal/fn.c

```diff
@@ -0,0 +1,26 @@
+#include <rlang.h>
+
+#include "internal.h"
+
+
+sexp* rlang_new_function(sexp* args, sexp* body, sexp* env) {
+  if (r_typeof(env) != r_type_environment) {
+    r_abort("`env` must be an environment");
+  }
+
+  args = KEEP(r_vec_coerce(args, r_type_pairlist));
+
+  sexp* node = args;
+  while (node != r_null) {
```

Comment:
Why do you prefer a `while` here instead of a `for`? I like that the `for` bounds the scope of the iterator variable, and separates the iteration from the action.